[ie/Bluesky] Add extractor

MellowKyler opened the pull request:
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Adds an extractor for bsky.app.

Fixes https://github.com/yt-dlp/yt-dlp/issues/10987

<details open><summary>Template</summary>

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

grqx commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,156 @@
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        handle = self._search_regex(self._VALID_URL, url, 'handle', group='handle')
```
```suggestion
        handle, video_id = self._match_valid_url(url).groups()
```

mikf commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,184 @@
+                   'depth': 6 if self.get_param('write_comments') else 0, 'parentHeight': 0},
+            expected_status=200)
+
+        formats, subs = self._extract_m3u8_formats_and_subtitles(
+            traverse_obj(meta, ('thread', 'post', 'embed', 'playlist')),
+            video_id, 'mp4', 'm3u8_native', m3u8_id='hls', fatal=False,
+            note='Downloading HD m3u8 information', errnote='Unable to download HD m3u8 information')
```
This does not extract the best available format.

For example for https://bsky.app/profile/blu3blue.bsky.social/post/3l4omssdl632g, the best format at 634x1080 resolution can be downloaded via
```
https://bsky.social/xrpc/com.atproto.sync.getBlob?did=did:plc:pzdr5ylumf7vmvwasrpr5bf2&cid=bafkreiamqpagnuwf72e5zj2xypoew3t2arxho4nse3le5l4zwou4hcrfne
```

See https://github.com/mikf/gallery-dl/commit/7d6520e15dba1a50e7c3609293efe3ddd521aa75 or https://github.com/danbooru/danbooru/pull/5816/commits/dc37dd19ee23bddab34f3d4b9c707b8dd898871b

MellowKyler replied:
Thanks! Implemented in 8c1dd6c4dabdabd2e8418d3dfcfc880e37a7a49c. :)

mikf replied:
I think you are supposed to extract *all* available formats for a yt-dlp extractor and not just the one with the best quality, so `_extract_m3u8_formats_and_subtitles` results plus blob.

MellowKyler replied:
Ah right, of course. Thank you! c17c3e5d52d6de14bb2d0960654a89449bb101ea

grqx commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,221 @@
+            'webpage_url': url,
+            'tags': post.get('labels', []) + traverse_obj(post, ('record', 'langs'), default=[]),
+            'comments': [] if not extractor else [*(extractor().get('comments'))[0]][1:],
+            'subtitles': self._merge_subtitles(subs),
```
```suggestion
            'subtitles': subs,
```

grqx commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,221 @@
+    }]
+
+    def _get_comments(self, meta):
+        yield self.traverse_replies(meta, traverse_obj(meta, ('post', 'uri'), default=''))
```
```suggestion
        yield from self.traverse_replies(meta, traverse_obj(meta, ('post', 'uri'), default=''))
```
Must be applied together with the suggestions on Line 219 and Line 198

grqx commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,221 @@
+            'comment_count': post.get('replyCount'),
+            'webpage_url': url,
+            'tags': post.get('labels', []) + traverse_obj(post, ('record', 'langs'), default=[]),
+            'comments': [] if not extractor else [*(extractor().get('comments'))[0]][1:],
```
```suggestion
            '__post_extractor': self.extract_comments(meta),
```
not sure if this is what you're trying to do(the suggestions on Line 157 and Line 198 must be applied to make this work)
Please also remember to update tests, if necessary

MellowKyler replied:
Thank you - this is much better :)

grqx commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,221 @@
+        handle = traverse_obj(post, ('author', 'handle'))
+        uploader = traverse_obj(post, ('author', 'displayName'))
+        description = traverse_obj(post, ('record', 'text'))
+        extractor = self.extract_comments(meta)
```
```suggestion
```
Must be applied together with the suggestions on Line 140 and Line 219

grqx commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,221 @@
+            'like_count': post.get('likeCount'),
+            'repost_count': post.get('repostCount'),
+            'comment_count': post.get('replyCount'),
+            'webpage_url': url,
```
```suggestion
```
this is automatically added

gavinbeatty commented:
This failed to get the cid correctly when I tested it today on a (private, so I shall not share it) bluesky post with an embedded video. As is, this branch resulted in an attempt to download a URL with `&cid=None`.

I haven't attempted to read `traverse_obj` or understand how it works, so I don't have a solution per se. But I inserted some debugging code to dump the `post` object. There was a reference to a m3u8 playlist in `post["embed"]["media"]["playlist"]`. No guarantees that naive traversal always works, but it did in my case.

Here's a copy of `post["embed"]` but anonymized:

```json
{"$type": "app.bsky.embed.recordWithMedia#view", "media": {"$type": "app.bsky.embed.video#view", "cid": "somecid", "playlist": "https://video.bsky.app/watch/did%3Aplc%3A<something>/somecid/playlist.m3u8", "thumbnail": "https://video.bsky.app/watch/did%3Aplc%3A<something>/somecid/thumbnail.jpg", "aspectRatio": {"height": 1080, "width": 1920}}
```

gavinbeatty commented:
Just noticed the playlist stuff already there. I'm not willing to spend the time going through the rigmarole necessary to fork and submit a PR to the bluesky branch, but the changes were simple:

1. traversals from embed -> playlist and embed -> thumbnail should have a middle "media" step, e.g., `traverse_obj(post, ('embed', 'media', 'playlist'))`
2. if `blob_cid` is falsey, don't append to `formats`

That worked fine for my single URL of interest.

mikf commented:
The correct CID can be obtained from
``` python
post["embed"]["video"]["ref"]["$link"] or
post["embed"]["video"]["cid"]
```

Sometimes there's an extra `media` in `embed`. In that case
``` python
post["embed"]["media"]["video"]["ref"]["$link"] or
post["embed"]["media"]["video"]["cid"]
```

This seems to work for gallery-dl, at least (https://github.com/mikf/gallery-dl/commit/4c391805de49c7b7f1b98357c2bd0096a5e88fce).

MellowKyler commented:
Updated to include the above paths (with the slight tweak that we go through "record" first, which gallery-dl also does). For example:
```python
post["record"]["embed"]["video"]["ref"]["$link"]
```
@gavinbeatty your example should work now if the post only has the m3u8 playlist. Thanks!

gavinbeatty commented:
Yep, your branch works for me now, thanks 👍

seproDev commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,223 @@
+            'upload_date': '20240921',
+            'description': 'OMG WE HAVE VIDEOS NOW',
+            'thumbnail': r're:https://video.bsky.app/watch/.*\.jpg$',
+            'alt_title': None,
```
You don't need to test for `None` values
```suggestion
```

seproDev commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,223 @@
+            'comment_count': int,
+            'webpage_url': 'https://bsky.app/profile/bsky.app/post/3l3vgf77uco2g',
+            'tags': 'count:2',
+            'subtitles': dict,
```
If you want to test for subtitles, you can test like so:
```suggestion
            'subtitles': {
                'en': 'mincount:1',
            },
```

seproDev commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,223 @@
+
+    def _real_extract(self, url):
+        handle, video_id = self._match_valid_url(url).groups()
+        did = handle if handle.startswith('did:') else self._download_json(
+            'https://public.api.bsky.app/xrpc/com.atproto.identity.resolveHandle',
+            video_id, query={'handle': handle}).get('did')
```
We don't need to request the `did` here, do we?
we can just call `getPostThread` with `f'at://{handle}/app.bsky.feed.post/{video_id}'`

For the `getBlob` request, we can use the did from the `getPostThread` response.

seproDev commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,223 @@
+                'uri': f'at://{did}/app.bsky.feed.post/{video_id}',
+                'depth': 1000 if getcomments else 0,
+                'parentHeight': 1000 if getcomments else 0,
+            }).get('thread')
```
Since the variable is mandatory
```suggestion
            })['thread']
```

seproDev commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,223 @@
+            }).get('thread')
+        post, record_embed = meta.get('post'), traverse_obj(meta, ('post', 'record', 'embed'))
+
+        formats, subs = self._extract_m3u8_formats_and_subtitles(
+            traverse_obj(post, ('embed', 'playlist'), ('embed', 'media', 'playlist')),
+            video_id, 'mp4', 'm3u8_native', m3u8_id='hls', fatal=False,
+            note='Downloading HD m3u8 information', errnote='Unable to download HD m3u8 information')
```
You should differentiate between the different kinds of post response. This will break for posts with embeds such as `https://bsky.app/profile/yunayuispink.bsky.social/post/3l7gqcfes742o`
You also don't need to define the complete path always.
```py
traverse_obj(post, ('embed', 'playlist'), ('embed', 'media', 'playlist'))
```
can be simplified to
```py
traverse_obj(post, ('embed', ('media', None), 'playlist'))
```

seproDev commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,223 @@
+            formats.append({
+                'format_id': 'blob',
+                'url': f'https://bsky.social/xrpc/com.atproto.sync.getBlob?did={did}&cid={blob_cid}',
+                'ext': mimetype2ext(traverse_obj(record_embed, ('video', 'mimeType')), 'mp4'),
+                'width': traverse_obj(record_embed, ('aspectRatio', 'width'), expected_type=int_or_none),
+                'height': traverse_obj(record_embed, ('aspectRatio', 'height'), expected_type=int_or_none),
+                'filesize': traverse_obj(record_embed, ('video', 'size'), expected_type=int_or_none),
```
You can combine `traverse_obj` calls. Please also do this for the info dict below.
```suggestion
                **traverse_obj(record_embed, {
                    'ext': ('video', 'mimeType', {mimetype2ext}),
                    'width': ('aspectRatio', 'width', {int_or_none}),
                    'height': ('aspectRatio', 'height', {int_or_none}),
                    'filesize': ('video', 'size', {int_or_none}),
                }),
```

seproDev reviewed (changes requested):
Initial review

seproDev commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,362 @@
+                        'filesize': ('video', 'size', {int_or_none}),
+                    }),
+                })
+            video_info = {
+                'formats': formats,
+                'subtitles': subs,
+                'thumbnail': url_or_none(quoted_media.get('thumbnail')),
+                'alt_title': quoted_embed.get('alt') or quoted_media.get('alt'),
+            }
+        else:
+            self.raise_no_formats('No video could be found in this post', expected=True)
+
+        handle = traverse_obj(post, ('author', 'handle'))
+        uploader = traverse_obj(post, ('author', 'displayName')) or handle
+
+        tags = traverse_obj(post, ('record', 'langs'), default=[])
+        if label_list := post.get('labels'):
+            tags.extend(label.get('val') for label in label_list)
+
+        return {
+            'id': video_id,
+            'title': f'{uploader} on Bluesky',
+            **video_info,
+            'uploader': uploader,
+            'channel': handle,
+            'uploader_id': did,
+            'channel_id': did,
+            'uploader_url': f'https://bsky.app/profile/{handle}',
+            'channel_url': f'https://bsky.app/profile/{did}',
+            'like_count': post.get('likeCount'),
+            'repost_count': post.get('repostCount'),
+            'comment_count': post.get('replyCount'),
+            'tags': tags,
+            'age_limit': 18 if {'sexual', 'porn', 'graphic-media'}.intersection(tags) else 0,
+            '__post_extractor': self.extract_comments(meta),
+            **traverse_obj(post, {
+                'timestamp': ('record', 'createdAt', {parse_iso8601}),
+                'description': ('record', 'text'),
+            }),
+        }
```
The remainder I would restructure to have a generic extraction method that can then be called with the separate sub paths:
```suggestion
        entries = []
        # app.bsky.embed.video.view/app.bsky.embed.external.view
        entries.extend(self._extract_videos(post, video_id))
        # app.bsky.embed.recordWithMedia.view
        entries.extend(self._extract_videos(
            post, video_id, embed_path=('embed', 'media'), record_subpath=('embed', 'media')))
        # app.bsky.embed.record.view
        if nested_post := traverse_obj(post, ('embed', 'record', ('record', None), {dict}, any)):
            entries.extend(self._extract_videos(
                nested_post, video_id, embed_path=('embeds', 0), record_path='value'))

        if len(entries) == 1:
            return entries[0]

        return self.playlist_result(entries, video_id)
```
I would also return both the embedded post and regular media of a post in a playlist if both are present.
This mirrors the twitter extractor.

seproDev commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,362 @@
+            if service.get('type') == 'AtprotoPersonalDataServer':
+                return service.get('serviceEndpoint')
+        return 'https://bsky.social'
+
```
For the video extract method, this is what I came up with:
```suggestion

    @staticmethod
    def _build_profile_url(path):
        return format_field(path, None, 'https://bsky.app/profile/%s', default=None)

    def _extract_videos(self, root, video_id, embed_path='embed', record_path='record', record_subpath='embed'):
        embed_path = variadic(embed_path, (str, bytes, dict, set))
        record_path = variadic(record_path, (str, bytes, dict, set))
        record_subpath = variadic(record_subpath, (str, bytes, dict, set))

        entries = []
        if external_uri := traverse_obj(root, (((*record_path, *record_subpath), embed_path), 'external', 'uri', {url_or_none}, any)):
            entries.append(self.url_result(external_uri))

        if playlist := traverse_obj(root, (*embed_path, 'playlist', {url_or_none})):
            formats, subtitles = self._extract_m3u8_formats_and_subtitles(
                playlist, video_id, 'mp4', m3u8_id='hls', fatal=False)
        else:
            return entries

        video_cid = traverse_obj(
            root, (*embed_path, 'cid', {str}),
            (*record_path, *record_subpath, 'video', 'ref', '$link', {str}))
        did = traverse_obj(root, ('author', 'did', {str}))
        if did and video_cid:
            endpoint = self.get_service_endpoint(did, video_id)
            formats.append({
                'format_id': 'blob',
                'url': update_url_query(self._BLOB_URL % endpoint, {'did': did, 'cid': video_cid}),
                **traverse_obj(root, (*embed_path, 'aspectRatio', {
                    'width': ('width', {int_or_none}),
                    'height': ('height', {int_or_none}),
                })),
                **traverse_obj(root, (*record_path, *record_subpath, 'video', {
                    'filesize': ('size', {int_or_none}),
                    'ext': ('mimeType', {mimetype2ext}),
                })),
            })
            for sub_data in traverse_obj(root, (
                    *record_path, *record_subpath, 'captions', lambda _, v: v['file']['ref']['$link'])):
                subtitles.setdefault(sub_data.get('lang') or 'und', []).append({
                    'url': update_url_query(self._BLOB_URL % endpoint, {'did': did, 'cid': sub_data['file']['ref']['$link']}),
                    'ext': traverse_obj(sub_data, ('file', 'mimeType', {mimetype2ext})),
                })

        entries.append({
            'id': video_id,
            'formats': formats,
            'subtitles': subtitles,
            **traverse_obj(root, {
                'id': ('uri', {lambda x: x.split('/')[-1] or None}, {str}),
                'thumbnail': (*embed_path, 'thumbnail', {url_or_none}),
                'alt_title': (*embed_path, 'alt', {str}),
                'uploader': ('author', 'displayName', {str}),
                'uploader_id': ('author', 'handle', {str}),
                'uploader_url': ('author', 'handle', {self._build_profile_url}),
                'channel_id': ('author', 'did', {str}),
                'channel_url': ('author', 'did', {self._build_profile_url}),
                'like_count': ('likeCount', {int_or_none}),
                'repost_count': ('repostCount', {int_or_none}),
                'comment_count': ('replyCount', {int_or_none}),
                'timestamp': ('indexedAt', {parse_iso8601}),
                'tags': ('labels', ..., 'val', {str}, all, {lambda x: list(set(x))}),
                'age_limit': ('labels', ..., 'val', {lambda x: 18 if x in ('sexual', 'porn', 'graphic-media') else None}, any),
                'description': (*record_path, 'text', {str}),
                'title': (*record_path, 'text', {lambda x: truncate_string(x, 50)}),
            }),
        })
        return entries

```
This requires the const `_BLOB_URL = '%s/xrpc/com.atproto.sync.getBlob'` to be added.

Some notes on the implementation/differences to yours:
- I tried to combine the different paths with paths to the record/embed being able to be passed
- I adjusted the uploader/channel(_id/_url) logic. I think like this it makes sense, and is somewhat in line with twitter.
- For the tags, I removed the `language`. Potentially, we could return this in the `language` field. I also deduplicated the tags.
- For the title, I think it makes more sense to go with the same value as the description. To not have the same issue as with the twitter extractor (potentially very long filenames that are not supported by the target filesystem) I would truncate the string.

seproDev commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,362 @@
+from .common import InfoExtractor
+from ..utils import int_or_none, mimetype2ext, parse_iso8601, traverse_obj, url_or_none
```
For my suggestions to work:
```suggestion
from ..utils import (
    format_field,
    int_or_none,
    mimetype2ext,
    parse_iso8601,
    truncate_string,
    update_url_query,
    url_or_none,
    variadic,
)
from ..utils.traversal import traverse_obj
```

seproDev reviewed (changes requested):
Sorry this took so long.

MellowKyler commented:
Thanks for the redesign @seproDev! I made a few modifications before committing:
- A raise_no_formats was added if no entries are found
- I modified the `traverse_obj` in `get_service_endpoint` (it was always returning the default before)
- A test was added for an at-uri (lovely touch!).
- Added a language field to the info dict. Because it's supposed to be a language code string and not a list of languages, I just grab the first one.

Happy to make whatever changes are needed to help you get this ready for the next release, and I'm the one who should apologize for this taking so long :)

seproDev commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,342 @@
+            url = f'https://plc.directory/{did}'
+        services = self._download_json(url, video_id, fatal=False)
+        return traverse_obj(
+            services, ('service', lambda _, x: x.get('type') == 'AtprotoPersonalDataServer',
```
Can be fatal inside of traversals
```suggestion
            services, ('service', lambda _, x: x['type'] == 'AtprotoPersonalDataServer',
```

seproDev commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,342 @@
+            'tags': [],
+            'language': 'en',
+        },
+    }]
```
Let's add one test with a playlist
```suggestion
    }, {
        'url': 'https://bsky.app/profile/purpleicetea.bsky.social/post/3l7gv55dc2o2w',
        'info_dict': {
            'id': '3l7gv55dc2o2w',
        },
        'playlist': [{
            'info_dict': {
                'id': '3l7gv55dc2o2w',
                'ext': 'mp4',
                'upload_date': '20241026',
                'description': 'One of my favorite videos',
                'comment_count': int,
                'uploader_url': 'https://bsky.app/profile/purpleicetea.bsky.social',
                'uploader': 'Purple.Ice.Tea',
                'thumbnail': 'https://video.bsky.app/watch/did%3Aplc%3Abjh5ffwya5f53dfy47dezuwx/bafkreicaldzyr6yr26ex4gyr6z2gbwvi53iolrcsqfipqqs3ieb2olcmiu/thumbnail.jpg',
                'channel_url': 'https://bsky.app/profile/did:plc:bjh5ffwya5f53dfy47dezuwx',
                'like_count': int,
                'channel_id': 'did:plc:bjh5ffwya5f53dfy47dezuwx',
                'repost_count': int,
                'timestamp': 1729973202,
                'tags': [],
                'uploader_id': 'purpleicetea.bsky.social',
                'title': 'One of my favorite videos',
            },
        }, {
            'info_dict': {
                'id': '3l77u64l7le2e',
                'ext': 'mp4',
                'title': 'hearing people on twitter say that bluesky isn\'...',
                'like_count': int,
                'uploader_id': 'thafnine.net',
                'uploader_url': 'https://bsky.app/profile/thafnine.net',
                'upload_date': '20241024',
                'channel_url': 'https://bsky.app/profile/did:plc:6ttyq36rhiyed7wu3ws7dmqj',
                'description': 'md5:dba7d54d6df7e5b79a24d7b287dafbe9',
                'tags': [],
                'alt_title': 'md5:9b1ee1937fb3d1a81e932f9ec14d560e',
                'uploader': 'T9',
                'channel_id': 'did:plc:6ttyq36rhiyed7wu3ws7dmqj',
                'thumbnail': 'https://video.bsky.app/watch/did%3Aplc%3A6ttyq36rhiyed7wu3ws7dmqj/bafkreih4xf5k22urq5kytjbnlj2djnzmlrqnvotnglmoy35mkpobrwtota/thumbnail.jpg',
                'timestamp': 1729731642,
                'comment_count': int,
                'repost_count': int,
            },
        }],
    }]
```

seproDev commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,342 @@
+                'age_limit': ('labels', ..., 'val', {lambda x: 18 if x in ('sexual', 'porn', 'graphic-media') else None}, any),
+                'description': (*record_path, 'text', {str}, any),
+                'title': (*record_path, 'text', {lambda x: truncate_string(x, 50)}),
+                'language': (*record_path, 'langs', ..., {str}, any),
```
After thinking on it, we should drop the language. This is supposed to be a format key specifying the language of the video/audio track. This is also used for `--embed-metadata`. Instead of overwriting the language, possibly incorrectly, I'd rather just not set it at all.
```suggestion
```

bashonly commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,379 @@
+                       'serviceEndpoint', {url_or_none}, any)) or 'https://bsky.social'
+
+    def _real_extract(self, url):
+        handle, video_id = self._match_valid_url(url).groups()
```
```suggestion
        handle, video_id = self._match_valid_url(url).group('handle', 'id')
```

bashonly commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,379 @@
+
+class BlueskyIE(InfoExtractor):
+    _VALID_URL = [
+        r'https?://(?:www\.)?(?:bsky\.app|main\.bsky\.dev)/profile/(?P<handle>[^/]+)/post/(?P<id>\w+)',
```
can we be more strict with handle matching? (maybe the `-` is even unnecessary, i don't see it in any tests)
```suggestion
        r'https?://(?:www\.)?(?:bsky\.app|main\.bsky\.dev)/profile/(?P<handle>[\w.:-]+)/post/(?P<id>\w+)',
```

or if not, then let's at least do this:
```suggestion
        r'https?://(?:www\.)?(?:bsky\.app|main\.bsky\.dev)/profile/(?P<handle>[^/?#]+)/post/(?P<id>\w+)',
```

MellowKyler replied:
Between the [handle](https://atproto.com/specs/handle) and [DID](https://atproto.com/specs/did) specs, valid characters are letters, digits, periods, underscores, colons, percent signs, and hyphens, so I'm thinking `(?P<handle>[\w.:%-]+)`.

bashonly replied:
edited my suggestions per your feedback

bashonly commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,379 @@
+class BlueskyIE(InfoExtractor):
+    _VALID_URL = [
+        r'https?://(?:www\.)?(?:bsky\.app|main\.bsky\.dev)/profile/(?P<handle>[^/]+)/post/(?P<id>\w+)',
+        r'at://(?P<handle>[^/]+)/app.bsky.feed.post/(?P<id>\w+)',
```
same question here
```suggestion
        r'at://(?P<handle>[\w.:-]+)/app.bsky.feed.post/(?P<id>\w+)',
```

bashonly commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,379 @@
+            },
+        }],
+    }]
+    _BLOB_URL = '%s/xrpc/com.atproto.sync.getBlob'
```
```suggestion
    _BLOB_URL_TMPL = '%s/xrpc/com.atproto.sync.getBlob'
```

bashonly commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,379 @@
+            endpoint = self.get_service_endpoint(did, video_id)
+            formats.append({
+                'format_id': 'blob',
+                'url': update_url_query(self._BLOB_URL % endpoint, {'did': did, 'cid': video_cid}),
```
```suggestion
                'url': update_url_query(
                    self._BLOB_URL_TMPL % endpoint, {'did': did, 'cid': video_cid}),
```

bashonly commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,379 @@
+            for sub_data in traverse_obj(root, (
+                    *record_path, *record_subpath, 'captions', lambda _, v: v['file']['ref']['$link'])):
+                subtitles.setdefault(sub_data.get('lang') or 'und', []).append({
+                    'url': update_url_query(self._BLOB_URL % endpoint, {'did': did, 'cid': sub_data['file']['ref']['$link']}),
```
line was too long before the `_BLOB_URL_TMPL` name change
```suggestion
                    'url': update_url_query(
                        self._BLOB_URL_TMPL % endpoint, {'did': did, 'cid': sub_data['file']['ref']['$link']}),
```

bashonly commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,379 @@
+            'formats': formats,
+            'subtitles': subtitles,
+            **traverse_obj(root, {
+                'id': ('uri', {lambda x: x.split('/')[-1] or None}, {str}),
```
```suggestion
                'id': ('uri', {lambda x: x.split('/')[-1]}, filter),
```

or maybe the `url_basename` util would be nice to use here?

bashonly replied:
btw you may need to merge master into this branch for this suggestion to work (`filter` is a new traverse_obj feature)

bashonly commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,379 @@
+                'repost_count': ('repostCount', {int_or_none}),
+                'comment_count': ('replyCount', {int_or_none}),
+                'timestamp': ('indexedAt', {parse_iso8601}),
+                'tags': ('labels', ..., 'val', {str}, all, {lambda x: list(set(x))}),
```
could use `orderedSet` from utils
```suggestion
                'tags': ('labels', ..., 'val', {str}, all, {orderedSet}),
```

bashonly commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,379 @@
+                'tags': ('labels', ..., 'val', {str}, all, {lambda x: list(set(x))}),
+                'age_limit': ('labels', ..., 'val', {lambda x: 18 if x in ('sexual', 'porn', 'graphic-media') else None}, any),
+                'description': (*record_path, 'text', {str}, any),
+                'title': (*record_path, 'text', {lambda x: truncate_string(x, 50)}),
```
note to self: `truncate_string` should have `partial_application`
```suggestion
                'title': (*record_path, 'text', {lambda x: truncate_string(x, 50)}),
```

bashonly commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,379 @@
+                'comment_count': ('replyCount', {int_or_none}),
+                'timestamp': ('indexedAt', {parse_iso8601}),
+                'tags': ('labels', ..., 'val', {str}, all, {lambda x: list(set(x))}),
+                'age_limit': ('labels', ..., 'val', {lambda x: 18 if x in ('sexual', 'porn', 'graphic-media') else None}, any),
```
line too long
```suggestion
                'age_limit': (
                    'labels', ..., 'val', {lambda x: 18 if x in ('sexual', 'porn', 'graphic-media') else None}, any),
```

bashonly commented on yt_dlp/extractor/bluesky.py:
```diff
@@ -0,0 +1,379 @@
+            (*record_path, *record_subpath, 'video', 'ref', '$link', {str}))
+        did = traverse_obj(root, ('author', 'did', {str}))
+        if did and video_cid:
+            endpoint = self.get_service_endpoint(did, video_id)
```
```suggestion
            endpoint = self._get_service_endpoint(did, video_id)
```

bashonly reviewed (changes requested):
in general I'd maybe add an empty line here or there to break up the larger methods into code blocks
